webhooks/core/service_mutator.go

```diff
@@ -0,0 +1,65 @@
+package core
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/webhook"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const (
+	apiPathMutateService = "/mutate-v1-service"
+)
+
+// NewServiceMutator returns a mutator for Service.
+func NewServiceMutator(lbClass string, logger logr.Logger) *serviceMutator {
+	return &serviceMutator{
+		logger:            logger,
+		loadBalancerClass: lbClass,
+	}
+}
+
+var _ webhook.Mutator = &serviceMutator{}
+
+type serviceMutator struct {
+	logger            logr.Logger
+	loadBalancerClass string
+}
+
+func (m *serviceMutator) Prototype(_ admission.Request) (runtime.Object, error) {
+	return &corev1.Service{}, nil
+}
+
+func (m *serviceMutator) MutateCreate(ctx context.Context, obj runtime.Object) (runtime.Object, error) {
+	svc := obj.(*corev1.Service)
+	if svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return svc, nil
+	}
+
+	if svc.Spec.LoadBalancerClass != nil && *svc.Spec.LoadBalancerClass != "" {
+		m.logger.Info("service already has loadBalancerClass, skipping", "service", svc.Name, "loadBalancerClass", *svc.Spec.LoadBalancerClass)
+		return svc, nil
+	}
+
+	svc.Spec.LoadBalancerClass = &m.loadBalancerClass
+	if svc.Annotations == nil {
+		svc.Annotations = make(map[string]string)
+	}
+	svc.Annotations["service.beta.kubernetes.io/aws-load-balancer-scheme"] = "internet-facing"
```

Comment:
I'm aware. But kubetest2 needs to confirm connectivity with the LB, so it would be a breaking change if we do not do this when the LB class is empty. One could very well say that kubernetes requires the CCM controller (and those that emulate it) to create a public cloud LB by default.

Reply:
What we probably have to do here is to introduce a compatibility mode flag. I can imagine people want both behaviors.

Reply:
I'm looking forward to merging the changes for v2.5.0. Lets limit the webhook to setting the spec.loadBalancerClass. We will leave the default service as internal for now. I'm open to providing a flag to change the default behavior in future releases.